Feat/6/TabBar UI 구현

SunnnySong opened the pull request:
![스크린샷 2023-07-15 오후 11 30 39](https://github.com/seunghyunCheon/LottoDiary/assets/85678496/cf33a9f4-0005-4b0d-95b2-3d24d15eea5b)

- TabBarComponents : TabBar에서 각 탭마다의 요소를 갖고 있는 타입
     ```swift
     private func configureViewControllers() {
        self.viewControllers = TabBarComponents.allCases.map { makeTabBarViewControllers($0) }
        selectedIndex = 1
    }
    ```  
    - TabBarController에서 각 탭의 NavigationController를 설정할 때, 코드 나열보단 map을 사용하는 것이 더 편리할 것이라 생각했다.
    - 만약 새로운 탭이 추가된다 해도 TabBarComponents의 요소만 추가하면 되기 때문에 변경에 용이하다.

- TabBarView : Custom한 UITabBar
- LottoQRButton : TabBar 가운데 영역에 있는 LottoQR 버튼
- TabBarController
     ```swift
     func tabBarController(_ tabBarController: UITabBarController, shouldSelect viewController: UIViewController) -> Bool {
        guard let selectedIndex = tabBarController.viewControllers?.firstIndex(of: viewController) else {
            return true
        }
        return selectedIndex == 2 ? false : true
    }    
    ```  
    - TabBar의 index가 2인 tabBarItem은 로또QR 버튼을 위한 빈 tabBarItem이 들어가 있다.
    - 해당 탭바는 단순히 로또QR 버튼 공간을 위해 넣어 놨기 때문에 해당 영역은 사용자 입력이 비활성화 되어야 한다.
    ```swift
    override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
        guard !clipsToBounds && !isHidden && alpha > 0 else { return nil }
        for member in subviews.reversed() {
            let subPoint = member.convert(point, from: self)
            guard let result = member.hitTest(subPoint, with: event) else { continue }
            return result
        }
        return nil
    }
    ````
    - 로또QR 버튼의 터치는 TabBarView에서 hitTest를 재정의해 해결했다.
    - TabBarView에서 subviews를 터치 이벤트 대상으로 지정했기 때문에 index==2인 영역은 터치를 비활성화 했지만 로또QR 버튼은 사용자 입력이 가능하다.

SunnnySong committed:
feat: TabBarComponents 타입 생성

- TabBarComponents 를 생성해 tabBarItem 마다의 title과 UIImage의 systemName을 관리한다.
- TabBar 확인을 위해 AppCoordinator의 start() 함수를 runMainFlow()로 변경했다.
- TabBarController에서 viewControllers 속성에 TabBarComponents들을 등록했다.

SunnnySong committed:
chore: TabBarController에서 extension으로 코드 분리

- TabBarController의 UITabBarControllerDelegate 프로토콜 채택 및 해당 함수 구현을 extension으로 코드를 분리했다.

SunnnySong committed:
feat: Custom TabBarView 생성

- custom한 TabBar를 위해 TabBarView를 생성했다.
- TabBarItem의 이미지와 TabBar 간의 간격이 너무 좁다고 느꼈기 때문에 TabBar의 height를 조정했다.
- TabBarView : tabBar의 배경 색상, tint 색상, 폰트를 설정했다.
- TabBarController에 TabBar를 setValue로 설정했다.
- index가 2인 TabBarItem 터치 비활성화 설정했다.

SunnnySong committed:
feat: LottoQRButton 생성 및 배치

- LottoQRButton 생성 : 이미지와 타이틀의 세로 배치를 위해 UIButton.Configuration 사용
- TabBarView에서 LottoQRButton 배치 : tabBar 영역을 벗어난 곳에서도 button의 터치를 인식하기 위해 hitTest override

SunnnySong committed:
feat: hitTest 관련 메서드 변경

- 전역변수로 선언한 lottoQRButton를 지역변수로 전환
- 기존 lottoQRButton의 frame에 속한 부분에만 터치가 되도록 hitTest 코드가 구현 -> 해당 TabBar의 subviews에 해당하는 모든 것들에 터치가 되도록 변경

SunnnySong committed:
refactor: UIButton의 이미지와 타이틀 세로 정렬

- 기존의 코드는 iOS15 이상부터 사용 가능한 UIButton.Configuration를 사용했다.
- iOS14 버전에 맞게 코드를 다시 구현했다.

SunnnySong committed:
chore: makeTabBarNavigationControllers 함수 이름 변경

SunnnySong committed:
chore: 생략 가능한 self 키워드 삭제

SunnnySong committed:
refactor: layoutSubviews -> init 코드로 변경

- configureTabBar 함수는 UITabBar의 정적인 속성들을 정의하고 있다.
- layoutSubviews은 frame에 대한 layout을 재정의할 때 더 적절한 메서드이고 정적 속성들은 init()에 넣는 것이 더 옳다고 생각해 변경

seunghyunCheon commented on LottoDairy/Support/Extension/UIButton+Extensions.swift:
```diff
@@ -0,0 +1,26 @@
+
+import UIKit
+
+extension UIButton {
+
+    func alignTextBelow(spacing: CGFloat = 4.0) {
+        guard let image = self.imageView?.image,
+              let titleLabel = self.titleLabel,
+              let titleText = titleLabel.text else {
+            return
+        }
+
+        let titleSize = titleText.size(withAttributes: [
+            NSAttributedString.Key.font: titleLabel.font as Any
+        ])
+
+        titleEdgeInsets = UIEdgeInsets(top: spacing, left: -image.size.width, bottom: -image.size.height, right: 0)
+        imageEdgeInsets = UIEdgeInsets(top: -(titleSize.height + spacing), left: 0, bottom: 0, right: -titleSize.width)
+    }
+}
```
`extension` 좋아요 👍
해당 `extension`도 추후 여러 곳에서 사용될 것 같아서 이 곳에 위치하는 게 좋아보이네요.

seunghyunCheon commented on LottoDairy/TabBarScene/Views/TabBarComponents.swift:
```diff
@@ -0,0 +1,50 @@
+        case .numbers:
+            return "번호 추첨"
+        case .empty:
+            return ""
```
empty case가 있는 이유가 무엇인가요?

SunnnySong replied:
LottoQR 버튼은 TabBar 영역 안에 있지만, TabBarItem이 아닌 UIButton으로 구현하고자 했습니다. 때문에 UIButton이 들어갈 빈 공간을 만들기 위해 empty case를 추가했어요!

seunghyunCheon commented on LottoDairy/TabBarScene/Views/TabBarView.swift:
```diff
@@ -0,0 +1,101 @@
+
+    private var shapeLayer: CALayer?
+
+    enum Constraints {
```
제약을 enum으로 정의한 부분 좋습니다!
해당 뷰에서만 사용한다면 `private`을 붙여도 좋을 것 같아요.

SunnnySong replied:
접근 제어자에 대한 코멘트 감사합니다! 해당 부분을 수정한 커밋입니다!
> dca9518e9b0bc7caab53a9c476c9f356e939e6d8

seunghyunCheon commented on LottoDairy/TabBarScene/Views/TabBarView.swift:
```diff
@@ -0,0 +1,101 @@
+    private func configureTabBar() {
+        barStyle = .black
+        tintColor = .designSystem(.white)
+        items?.forEach { $0.setTitleTextAttributes(
+            [.font : UIFont.gmarketSans(size: .caption, weight: .medium)],
+            for: .normal)
+        }
```
취향에 따라 다르겠지만 코드가 조금 길어진다면 클로저 축약문법 대신에 요소 이름을 쓰는 것도 좋을 것 같아요!
```swift
items?.forEach { tabbarItem in
      tabbarItem.setTitleTextAttributes(
          [.font: UIFont.gmarketSans(size: .caption, weight: .medium)],
          for: .normal
      )
}
```

SunnnySong replied:
오 축약 문법은 코드 길이가 짧아지는 대신 코드의 의미를 이해하는데 조금 어려울 것 같긴 한것 같습니다! 
축약 문법 수정한 커밋은 아래와 같습니다~
> 0b12564aa034213298bb69786e76bf516e2ea236

seunghyunCheon commented on LottoDairy/TabBarScene/Views/TabBarView.swift:
```diff
@@ -0,0 +1,101 @@
+        configureLottoQRButton()
+    }
+
+    override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
+        guard !clipsToBounds && !isHidden && alpha > 0 else { return nil }
+        for member in subviews.reversed() {
+            let subPoint = member.convert(point, from: self)
+            guard let result = member.hitTest(subPoint, with: event) else { continue }
+            return result
+        }
+        return nil
+    }
```
`hitTest` 재정의 좋아요 👍

seunghyunCheon commented on LottoDairy/TabBarScene/TabBarController.swift:
```diff
@@ -16,30 +16,50 @@ final class TabBarController: UITabBarController, UITabBarControllerDelegate, Ta
         }
     }
+    
+    func tabBarController(_ tabBarController: UITabBarController, shouldSelect viewController: UIViewController) -> Bool {
+        guard let selectedIndex = tabBarController.viewControllers?.firstIndex(of: viewController) else {
+            return true
+        }
+        return selectedIndex == 2 ? false : true
+    }
+
```
tabbarView의 hitTest재정의를 통해 사용자 input을 받을 수 있다는 것 까지는 이해했습니다. 그런데 QR코드 화면을 보여주는 코드는 추후 어디에 작성하실 생각이신가요?

SunnnySong replied:
해당 코드는 TabBarController에 구현할 예정입니다! 기존 TabBarItem에 UINavigationController를 연결한 코드가 TabBarController에 있기 때문에 해당 코드 또한 이 파일 안에 들어가 있어야 한다고 생각합니다.
lottoQRButton에 추후 추가될 UIControl + Combine의 publisher(for:) 함수를 사용해 화면 전환에 관한 코드를 구현할 것입니다.

SunnnySong committed:
review: Constraints enum에 private 설정

SunnnySong committed:
review: TabBarView의 축약 문법 수정

SunnnySong committed:
feat: LottoQRController 연결

- LottoQRViewController 생성
- LottoQRFlowProtocol 생성
    - 추후 LottoQRFlow에서 넘어갈 화면을 위해 FlowProtocol을 구현할 예정. 지금은 빈 프로토콜로 구현.
- LottoQRModuleFactory protocol 생성
- ModuleFactoryImp에 makeLottoQRFlow() 추가
- CoordinatorFactoryImp에 makeLottoQRCoordinator() 추가
